snuba/utils/streams/batching.py

```diff
@@ -61,6 +62,150 @@ class Offsets:
     hi: int  # exclusive
 
 
+@dataclass
+class Batch(Generic[TResult]):
+    results: MutableSequence[TResult] = field(default_factory=list)
+    offsets: MutableMapping[Partition, Offsets] = field(default_factory=dict)
+    created: float = field(default_factory=lambda: time.time())
+    messages_processed_count: int = 0
+    # the total amount of time, in milliseconds, that it took to process
+    # the messages in this batch (does not included time spent waiting for
+    # new messages)
+    processing_time_ms: float = 0.0
+
+
+class BatchProcessor:
+    """
+    The ``BatchProcessor`` is a message processor that accumulates processed
+    message values, periodically flushing them after a given duration of time
+    has passed or number of output values have been accumulated.
+    """
+
+    def __init__(
+        self,
+        consumer: Consumer[TPayload],
+        worker: AbstractBatchWorker[TPayload, TResult],
+        max_batch_size: int,
+        max_batch_time: int,
+        metrics: MetricsBackend,
+    ) -> None:
+        self.__consumer = consumer
```

Comment:
I think this creates some tight coupling between this class and the kafka consumer that may not be needed. 
Did you mean this in the description with

> Define any sort of strategy for the processor to communicate offsets back to the consumer.

?

If the BatchingConsumer (or how it will be called) owns the lifecycle of the consumer and of the assignment, we could restrict the management of the kafka consumer there. Leaking out the consumer allows the BatchingProcessor to potentially drive that lifecycle.

It seems that what this class need to do is really having a way to asynchronously communicating back that the processing of all messages up to a given offset have bee\n completed without caring of what actually that means for the BatchingConsumer.
What about passing a callback from the BatchingConsumer to the BatchProcessor. The BatchProcessor calls this funnction/protocol/.... signaling the last completed offset. The callback itself commits.
It would also make it easier to unit test BatchProcessor.

Or am I missing some dependency between BatchProcessor and the consumer that is not yet visible at this point ?

Reply:
> Or am I missing some dependency between BatchProcessor and the consumer that is not yet visible at this point ?

No dependency (at least that I've considered.) I agree this is not good design and yes, that is what I was trying to describe in the description as a (somewhat intentional) problem.

I think the callback approach is a reasonable one (and was kind of what I was thinking as well, so good to hear it mentioned as well.)

I'd prefer to do this in a follow up just to minimize the amount of stuff going on in this change. I'm happy to make that a ticket to ensure it doesn't get lost, though I think it'll probably make its way into the follow up to this change (referenced earlier) that cleans up the abstractions more, especially as the unit testing aspect is a good point. If you feel strongly that this needs to be done sooner rather than later I can include it in this change, though, or even merge the follow ups into this change (though it will start becoming quite large.)

Reply:
Added this here: c393989e301f04071762d6d62479a2da44cb0c77